galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt

```diff
@@ -15,6 +15,2538 @@
+   */
+  abstract fun copyRecord(f: Int, t: Int)
+
+  /**
+   * Returns the containing block.
+   */
+  fun getForm(): VForm = block!!.getForm()
+
+  /**
+   * Returns true if field is never displayed.
+   */
+  fun isInternal(): Boolean {
+    return access[VConstants.MOD_QUERY] == VConstants.ACS_HIDDEN
+            && access[VConstants.MOD_INSERT] == VConstants.ACS_HIDDEN
+            && access[VConstants.MOD_UPDATE] == VConstants.ACS_HIDDEN
+  }
+
+  // ----------------------------------------------------------------------
+  // PROTECTED UTILS
+  // ----------------------------------------------------------------------
+  /**
+   * trails (backups) the record if called in a transaction and restore it
+   * if the transaction is aborted.
+   */
+  protected fun trail(r: Int) {
+    if (!isTransient() && !hasTrigger(VConstants.TRG_VALUE) && alias == null) {
+      block!!.trailRecord(r)
+    }
+  }
+
+  /**
+   * Marks the field changed, trails the record if necessary
+   */
+  protected fun setChanged(r: Int) {
+    if (!isTransient() && !hasTrigger(VConstants.TRG_VALUE) && alias == null) {
+      block!!.setRecordChanged(r, true)
+    } else {
+      block!!.updateAccess(r)
+    }
+    changed = true
+    changedUI = false
+    fireValueChanged(r)
+  }
+
+
+
+
+  /**
+   * Marks the field changed, trails the record if necessary
+   */
+  fun setChanged(changed: Boolean) {
+    if (changed && block!!.getActiveRecord() != -1) {
+      block!!.setRecordChanged(block!!.getActiveRecord(), true)
+    }
+    this.changed = changed
+  }
+
+  /**
+   * Checks that field value exists in list
+   */
+  private fun checkList() {
+    if (!getForm().forceCheckList()) {
+      // Oracle doesn't force the value to be in the list
+      return
+    }
+    val SELECT_IS_IN_LIST = " SELECT   1                                      " +
+            " FROM     $2                                     " +
+            " WHERE    $1 = $3"
+
+    val SELECT_MATCHING_STRINGS = " SELECT   $1                                     " +
+            " FROM     $2                                     " +
+            " WHERE    {fn SUBSTRING($1, 1, {fn LENGTH(#3)})} = #3    " +
+            " ORDER BY 1"
+
+    if (isNull(block!!.getActiveRecord())) {
+      return
+    }
+    if (list == null) {
+      return
+    }
+    val alreadyProtected: Boolean = getForm().inTransaction()
+    if (this !is VStringField) {
+      var exists: Boolean
+
+      try {
+        while (true) {
+          try {
+            if (!alreadyProtected) {
+              getForm().startProtected(null)
+            }
+            val query = Query(getForm().dBContext.defaultConnection)
+
+            query.addString(list!!.getColumn(0).column)
+            query.addString(evalListTable())
+            query.addString(getSql(block!!.getActiveRecord()))
+            query.open(SELECT_IS_IN_LIST)
+            exists = query.next()
+            query.close()
```

Comment:
you won't need to introduce a new variable query if you use let  : 
Query(getForm().dBContext.defaultConnection).let {
              it.addString(list!!.getColumn(0).column)
              it.addString(evalListTable())
              it.addString(getSql(block!!.getActiveRecord()))
              it.open(SELECT_IS_IN_LIST)
              exists = it.next()
              it.close()
            }